### PR TITLE
Downloading videos using Background Fetch API

### LINKED ISSUE #25: Downloading static video to offline storage via Background Fetch API
## Feature description

As a user, I should be able to download a static video to offline storage via background fetch API

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

Design note and AC: Presentation of file being downloaded is different between background fetch and fetch API. On background fetch there is a banner on system that says being downloaded, will happen regardless of leaving browser. Fetch API handled by application, need indicator built into application to expose progress to user. 

They also should be aware of where their downloads are going.

Relates to #24, #25, #26, #35

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by dero**
## Short Screencast

https://user-images.githubusercontent.com/433570/142623506-e4557a8c-9ab4-4706-9aa9-a215161ca81d.mp4

## Summary

This PR adds support for downloading single videos and streamed videos using Background Fetch API.

We're adding an `offline` version of MPEG DASH manifests to all streamed videos. This `offline` manifest only contains one selected video and audio representation (HD H.264 MP4), which allows us to only download this one representation for offline playback.

We've been doing this on-the-fly in the past, but with the introduction of Background Fetch API, it arguably make more sense to move towards using the more explicit separate offline manifests, because it would bring a larger amount of complexity to the codebase if we tried to manipulate DASH manifest data on the fly in the service worker.

* The `DOMParser` object is not available in the worker context, meaning that parsing the manifest file would have to be done by an external library like [xmldom](https://www.npmjs.com/package/xmldom). (Extra bundle weight, one more thing to maintain and rely on.)
* We can't instantiate a `video` element or use its `canPlayType` method statically from the worker context. That means we can't test which of the representations in the manifest are actually playable on the current device and in the current browser. We would need to build a cache layer in IDB in order to figure that out ahead of time and cache it for later when SW is invoked with the downloaded files.

We can avoid both issues by uploading the offline version of each manifest upfront. Plus this approach allows us to simplify the code quite a bit.

The main downside, I think, is that users will not be able to select their preferred download resolution with static offline manifests in place. Note: This is not a regression, we never had a setting for download quality. But this change does not allow for such a setting should we decide to implement it in the future.

**Edit:** Thinking about this a bit more, maybe it would be possible to leverage the Media Capabilities API to detect support and gain playback performance estimates for individual resources listed in DASH manifests – the API should be available from worker threads.

We might decide to pick this back up later.

## Questions

- Should the "Download videos in the background" setting be enabled by default?
- Do we create a new video post for "Background Fetch API"?

Fixes #25

### CODE REVIEW

File: src/js/classes/BackgroundFetch.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      // eslint-disable-next-line compat/compat
+      navigator.serviceWorker.ready.then(async (swReg) => {
+        this.maybeAbort(swReg);
+
+        /** @type {BackgroundFetchRegistration} */
+        const bgFetch = await swReg.backgroundFetch.fetch(
+          this.id,
+          urls,
+          {
+            title: videoData.title,
```
Comment (beaufortfrancois) [inline src/js/classes/BackgroundFetch.js:56]: `videoData.title` may end up being confusing. See video at http://localhost:5000/multiple-sources/ for instance.
Shall we have something like `Downloading "${videoData.title}" video`? ([Source](https://github.com/WICG/background-fetch#:~:text=title%20%2D%20Something%20descriptive%20to%20show%20in%20UI%2C%20such%20as%20%22Uploading%20%27Holiday%20in%20Rome%27%22%20or%20%22Downloading%20%27Catastrophe%20season%202%20episode%201%27%22.))

image

File: src/js/classes/BackgroundFetch.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    Promise.all(requests).then((responses) => {
+      const sizes = responses.map((response) => response.headers.get('Content-Length'));
+      const downloadTotal = sizes.includes(null)
+        ? null
+        : sizes.reduce((total, size) => total + parseInt(size, 10), 0);
+
+      // eslint-disable-next-line compat/compat
+      navigator.serviceWorker.ready.then(async (swReg) => {
+        this.maybeAbort(swReg);
```
Comment (beaufortfrancois) [inline src/js/classes/BackgroundFetch.js:49]: I guess you'd want to await there before or is this intentional?
```suggestion
        await this.maybeAbort(swReg);
```

File: src/js/sw/sw.js [hunk 3]
```diff
@@ -217,3 +220,53 @@ const fetchHandler = async (event) => {
 self.addEventListener('install', precacheAssets);
 self.addEventListener('activate', clearOldCaches);
 self.addEventListener('fetch', fetchHandler);
+
+/** @type {MessagePort} */
+let messageChannelPort;
```
Comment (beaufortfrancois) [inline src/js/sw/sw.js:225]: Some quick comment to explain where it comes from may be helpful to readers.

File: src/js/sw/sw.js [hunk 4]
```diff
@@ -217,3 +220,53 @@ const fetchHandler = async (event) => {
...
+    if (event.data.type === 'channel-port') {
+      [messageChannelPort] = event.ports;
+    }
+  },
+);
+
+const bgFetchHandler = async (e) => {
+  /** @type {BackgroundFetchRegistration} */
+  const bgFetchRegistration = e.registration;
+  const bgFetch = new BackgroundFetch();
```
Comment (beaufortfrancois) [inline src/js/sw/sw.js:239]: Nit: Shall we move this until we actually need `videoId`?

```js

const bgFetchHandler = async (e) => {
  /** @type {BackgroundFetchRegistration} */
  const bgFetchRegistration = e.registration;
  const records = await bgFetchRegistration.matchAll();
  const urls = records.map((record) => record.request.url);
  
  if (urls.length === 0)
    return;
  
  const responsePromises = records.map((record) => record.responseReady);
  const responses = await Promise.all(responsePromises);
    
  const bgFetch = new BackgroundFetch();
  bgFetch.fromRegistration(bgFetchRegistration);
...
```

File: src/js/sw/sw.js [hunk 5]
```diff
@@ -217,3 +220,53 @@ const fetchHandler = async (event) => {
 self.addEventListener('install', precacheAssets);
 self.addEventListener('activate', clearOldCaches);
 self.addEventListener('fetch', fetchHandler);
+
```
Comment (beaufortfrancois) [inline src/js/sw/sw.js:223]: Shall we gate all the code added only `if (!('BackgroundFetchManager' in self))`?

**Review by beaufortfrancois — COMMENTED** [review]
Here's my first pass.
Thanks @dero for working on it!

File: src/js/sw/sw.js (hunk 5 above)
Comment (dero) [reply src/js/sw/sw.js:223]: @beaufortfrancois Thank you, addressed in 7979e32.

File: src/js/sw/sw.js (hunk 4 above)
Comment (dero) [reply src/js/sw/sw.js]: @beaufortfrancois Good suggestion, thanks! Addressed in 7979e32.

File: src/js/sw/sw.js (hunk 3 above)
Comment (dero) [reply src/js/sw/sw.js]: Added in 7979e32.

File: src/js/classes/BackgroundFetch.js (hunk 2 above)
Comment (dero) [reply src/js/classes/BackgroundFetch.js]: Added in 7979e32.

File: src/js/classes/BackgroundFetch.js (hunk 1 above)
Comment (dero) [reply src/js/classes/BackgroundFetch.js]: Updated in 7979e32.

**Comment by beaufortfrancois** [comment]
LGTM. Thanks @dero!

**Review by derekherman — APPROVED** [review]

**Comment by Swansoffie02** [comment]
> ## Short Screencast
>  kino-background-fetch-api.mp4 
> ## Summary
> This PR adds support for downloading single videos and streamed videos using Background Fetch API.
> 
> We're adding an `offline` version of MPEG DASH manifests to all streamed videos. This `offline` manifest only contains one selected video and audio representation (HD H.264 MP4), which allows us to only download this one representation for offline playback.
> 
> We've been doing this on-the-fly in the past, but with the introduction of Background Fetch API, it arguably make more sense to move towards using the more explicit separate offline manifests, because it would bring a larger amount of complexity to the codebase if we tried to manipulate DASH manifest data on the fly in the service worker.
> 
> * The `DOMParser` object is not available in the worker context, meaning that parsing the manifest file would have to be done by an external library like [xmldom](https://www.npmjs.com/package/xmldom). (Extra bundle weight, one more thing to maintain and rely on.)
> * We can't instantiate a `video` element or use its `canPlayType` method statically from the worker context. That means we can't test which of the representations in the manifest are actually playable on the current device and in the current browser. We would need to build a cache layer in IDB in order to figure that out ahead of time and cache it for later when SW is invoked with the downloaded files.
> 
> We can avoid both issues by uploading the offline version of each manifest upfront. Plus this approach allows us to simplify the code quite a bit.
> 
> The main downside, I think, is that users will not be able to select their preferred download resolution with static offline manifests in place. Note: This is not a regression, we never had a setting for download quality. But this change does not allow for such a setting should we decide to implement it in the future.
> 
> **Edit:** Thinking about this a bit more, maybe it would be possible to leverage the Media Capabilities API to detect support and gain playback performance estimates for individual resources listed in DASH manifests – the API should be available from worker threads.
> 
> We might decide to pick this back up later.
> 
> ## Questions
> * Should the "Download videos in the background" setting be enabled by default?
> * Do we create a new video post for "Background Fetch API"?
> 
> Fixes #25

**Comment by lkraav** [comment]
Why the revert?

**Pull request merged**